src/config.ts

```diff
@@ -1,10 +1,271 @@
+  const validator = new ajv();
+  if (!validator.validate(schema, rawConfig)) {
+    logger.error(`Your config file at ${getConfigPath()} is not valid.`);
+    logger.debug('You can specify the path with the `QCS_CONFIG` or `QC` environment variable.');
+    logger.debug(JSON.stringify(validator.errors, undefined, 2));
+    process.exit(1);
+  }
+};
+
+const logCredentialUpdateInstructions = (qcsURL: string) => {
+  logger.debug(`Please visit ${qcsURL}/auth/token to update your client credentials.`);
+  let authPath = '~/.qcs/user_auth_token';
+  if (forestEnvironment) {
+    authPath = `~/.qcs/user_auth_token__${forestEnvironment}`;
+  }
+  logger.debug(`Save your credentials to \`${authPath}\`.`);
+  logger.debug(
+    'You may also use the `AUTH` environment variable to customize the path to your credentials.');
+};
+
+export interface RawQCSConfig {
+  user_id?: string;
+  qcs_url?: string;
+  url: string;
+  [key: string]: string | undefined;
+}
+
+export class QCSConfig {
+  protected readonly rawConfig: RawQCSConfig;
+  protected readonly rawConfigSchema: any = QCS_CONFIG_SCHEMA;
+  protected qmiAccessToken?: AccessToken;
+  protected userAccessToken?: AccessToken;
+  private nAuthRefreshes = 0;
+
+  constructor() {
+    this.userAccessToken = getAccessToken('user');
+    this.qmiAccessToken = getAccessToken('qmi');
+    this.rawConfig = getRawConfig();
+    if (forestEnvironment === 'test' && !this.rawConfig.url) {
+      // tslint:disable-next-line
+      this.rawConfig.url = 'http://localhost:8000';
+      this.rawConfig.forest_admin_key = 'pkey';
+    }
+    this.assertValidConfig();
+    this.assertCredentials();
+  }
+
+  get url() {
+    return this.rawConfig.url;
+  }
+
+  get qcsURL() {
+    if (this.rawConfig.qcs_url) {
+      return this.rawConfig.qcs_url;
+    }
+    if (/:8000/.test(this.rawConfig.url)) {
+      return this.rawConfig.url.replace(/:8000/, ':3000');
+    }
+    return this.rawConfig.url.replace(/forest-server\./, '');
+  }
+
+  get headers() {
+    const headers: { [key: string]: string } = {};
+    if (this.userAccessToken) {
+      headers.Authorization = `Bearer ${this.userAccessToken.access_token}`;
+    } else if (this.qmiAccessToken) {
+      headers['X-QMI-AUTH-TOKEN'] = this.qmiAccessToken.access_token;
+    }
+    if (this.rawConfig.user_id) {
+      // WARN: This will be deprecated.
+      headers['X-User-Id'] = this.rawConfig.user_id;
+    }
+    return headers;
+  }
+
+  // request will retry any 401 or 403 response after refreshing
+  // the available user or QMI access token.
+  public async request<T>(opts: request.Options): Promise<T>{
+    try {
+      return this._requestWithDefaults(opts);
+    } catch (err) {
+      if (err.response && [401, 403].indexOf(err.response.statusCode) >= 0) {
+        await this.refreshAuthToken();
+        return this._requestWithDefaults(opts);
+      }
+      throw err;
+    }
+  }
+
+  public async refreshAuthToken() {
+    this.nAuthRefreshes += 1;
```

Comment:
Should this reset to 0 when it succeeds? Or is the object ephemeral?

Reply:
Well, it should be practically impossible for the token to expire over the course of a single CLI command.